docs/pallas/tpu/core_map.md

```diff
@@ -0,0 +1,358 @@
+```
+
+    Running with 4 TPU v5 devices.
+
+
+In addition to the typical TPU device mesh, you need to make a mesh of cores. Consider this as an addition dimension called "core", with length 2, in addition to the 4-device mesh you work with. That is 8 cores in total.
+
+
+```python
+# Mesh of devices
+mesh = jax.make_mesh((jax.device_count(),), ('device',))
+print(mesh)
+
+# Mesh of cores, within a JAX device
+tc_mesh = pltpu.create_tensorcore_mesh('core')
+print(tc_mesh)
+
+num_devices = mesh.size
+num_cores = len(tc_mesh.devices)
+print(f"There are {num_devices} devices, and {num_cores} cores each.")
+```
+
+    Mesh('device': 4, axis_types=(Auto,))
+    TensorCoreMesh(devices=array([TensorCore(id=0), TensorCore(id=1)], dtype=object), axis_names=('core',))
+    There are 4 devices, and 2 cores each.
+
+
+## A simple per-core kernel
+
+`pl.core_map` allows you to write per-core local code, just as `jax.shard_map` allows you to write per-device code.
+
+In the example kernel below, each core has its own VMEM and semaphore allocations. As with normal kernel, you can initiate copy between HBM and VMEM refs using `async_copy`.
+
+**Communication amongst cores**
+
+Before making a inter-core communication, you may need to do a global barrier signal (`pltpu.semaphore_signal`), to make sure all the destination semaphores have been properly initialized.
+
+After that, use `pltpu.make_async_remote_copy` to send the actual data. The `device_id` allows you to specify the destination using only the axis coordinate(s) that are different from the source.
+
+
+
+```python
+# This runs on every core
+def swap_cores_kernel(in_hbm, out_hbm,
+                      in_vmem, scratch_vmem, out_vmem,
+                      sem, send_sem, recv_sem):
+  core_index = jax.lax.axis_index('core')
+  num_cores = jax.lax.axis_size('core')
+  slc_size = in_hbm.shape[-1] // num_cores
+  slc = pl.ds(core_index * slc_size, slc_size)
+
+  # A barrier to make sure all cores have entered run_scoped.
+  # You won't need this if not doing inter-core communications.
+  sem0 = pltpu.get_barrier_semaphore()
+  for i in range(num_devices):
+    for j in range(num_cores):
+      pltpu.semaphore_signal(sem0, 1, device_id={'device': i, 'core': j})
+  pltpu.semaphore_wait(sem0, num_devices * num_cores)
+
+  # Copy in the input
+  pltpu.async_copy(in_hbm.at[:, slc], in_vmem, sem).wait()
+
+  # Swap data between core 0 and core 1
+  dst_core = (core_index + 1) % num_cores
+  the_copy = pltpu.make_async_remote_copy(
+      in_vmem, scratch_vmem, send_sem, recv_sem, device_id={'core': dst_core},
+  )
+  the_copy.start()
+  the_copy.wait()
+
+  # Core-local compute
+  out_vmem[...] = scratch_vmem[...] * 2
+
+  # Copy out the output
+  pltpu.async_copy(out_vmem, out_hbm.at[:, slc], sem).wait()
+
+```
+
+Once you have the local kernel:
+
+ * Wrap it with `pl.run_scoped` and HBM refs, so that the required scratch spaces (VMEM and semaphores) are allocated, one copy per core.
```

Comment:
Allocate per-core scratch spaces (VMEM and semaphores) using `run_scoped`, then run it with HBM refs and the allocations,